feat: Add concat operation to constraint streams

Christopher-Chianelli opened the pull request:
The concat operation acts as an or between two constraint streams, propogating both stream tuples' downstream. It appears update/retract may be called before an insert, meaning the out tuple stored inside the in tuple is not set. As such, for update, if the out tuple is not set, it does an insert instead, and for retract, it simply ignores the operation.

Christopher-Chianelli committed:
feat: Add concat operation to constraint streams

The concat operation acts as an or between two constraint streams,
propogating both stream tuples' downstream. It appears update/retract
may be called before an insert, meaning the out tuple stored inside
the in tuple is not set. As such, for update, if the out tuple is
not set, it does an insert instead, and for retract, it simply
ignores the operation.

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/BavetConstraintSessionFactory.java:
```diff
@@ -140,6 +141,13 @@ private long determineLayerIndex(AbstractNode node, NodeBuildHelper<Score_> buil
             var rightParentNode = buildHelper.findParentNode(rightParent);
             return Math.max(leftParentNode.getLayerIndex(), rightParentNode.getLayerIndex()) + 1;
+        } else if (node instanceof AbstractConcatNode<?> concatNode) {
+            var nodeCreator = (BavetJoinConstraintStream<?>) buildHelper.getNodeCreatingStream(concatNode);
```
Why would concat implement join? The two operations have barely anything in common.

Christopher-Chianelli replied:
Largely for a fact that like `join`, `concat` has two parents, and the only two methods on `BavetJoinConstraintStream` are getters for the two parents. IMO, it would be pointless for two classes to have exactly the same methods, although those two methods could be extracted into a super interface `BavetBinaryOperatorResultStream` and have `BavetJoinConstraintStream` extend that (essentially making `BavetJoinConstraintStream` a marker interface).

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetAbstractBiConstraintStream.java:
```diff
@@ -359,6 +359,19 @@ public BiConstraintStream<A, B> distinct() {
+        var rightBridge = new BavetForeBridgeBiConstraintStream<>(constraintFactory, other);
+        var concatStream = new BavetConcatBiConstraintStream<>(constraintFactory, leftBridge, rightBridge);
+        return constraintFactory.share(concatStream, concatStream_ -> {
+            // Connect the bridges upstream, as it is an actual new join.
```
This is not a join; let's not call it like that.

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetBiConcatNode.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    protected BiTuple<A, B> getOutTuple(BiTuple<A, B> inTuple) {
+        return new BiTuple<>(inTuple.factA, inTuple.factB, outputStoreSize);
```
Since concat does not modify the tuples, it doesn't need to create new ones.
I'm betting it will be much more efficient if it's rewritten as a pass-through node, without node creation.

Christopher-Chianelli replied:
Tried that, the queue did not like it since it sees the inTuple was already inserted.

Christopher-Chianelli replied:
Turn out we do need new tuples, since each node can only have 1 tuple source, and since concat has two parents, it must be a tuple source, and a tuple source must produce new tuples, not pass through existing ones.

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractConcatNode.java:
```diff
@@ -0,0 +1,103 @@
+
+        TupleState state = outTuple.state;
+        if (!state.isActive()) {
+            // Impossible because they shouldn't linger in the indexes.
```
There aren't any indexes.

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractConcatNode.java:
```diff
@@ -0,0 +1,103 @@
+public abstract class AbstractConcatNode<Tuple_ extends AbstractTuple>
+        extends AbstractNode
+        implements LeftTupleLifecycle<Tuple_>, RightTupleLifecycle<Tuple_> {
+    private final int inputStoreIndexLeftOutTupleList;
+    private final int inputStoreIndexRightOutTupleList;
```
These vars need better naming, as there are no indexes and no out tuple list.

triceo reviewed (commented):
Nice! 
I think we can still make the node itself more efficient.

Also, some minor comments. (Please make sure to apply them consistently in all cardinalities of streams.)

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractConcatNode.java:
```diff
@@ -0,0 +1,103 @@
+    @Override
+    public final void insertLeft(Tuple_ tuple) {
+        Tuple_ outTuple = getOutTuple(tuple);
+        tuple.setStore(inputStoreIndexLeftOutTupleList, outTuple);
```
If you store the node-specific tuple state in the store instead of the outtuple, you won't need to create the outtuple just to store the state.

Christopher-Chianelli committed:
docs: Add explanation for why concat cannot be pass-through in javadocs

Christopher-Chianelli committed:
test: Add reference equal groupby test for concat

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/uni/BavetConcatUniConstraintStream.java:
```diff
@@ -0,0 +1,104 @@
+        /*
+         * Bridge streams do not implement equality because their equals() would have to point back to this stream,
+         * resulting in StackOverflowError.
+         * Therefore we need to check bridge parents to see where this join node comes from.
```
Not a join node.
(Same comment also likely in Bi, Tri and Quad.)

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/BavetConstraintSessionFactory.java:
```diff
@@ -140,6 +142,13 @@ private long determineLayerIndex(AbstractNode node, NodeBuildHelper<Score_> buil
+            var nodeCreator = (BavetConcatConstraintStream<?>) buildHelper.getNodeCreatingStream(concatNode);
+            var leftParent = nodeCreator.getLeftParent();
+            var rightParent = nodeCreator.getRightParent();
+            var leftParentNode = buildHelper.findParentNode(leftParent);
+            var rightParentNode = buildHelper.findParentNode(rightParent);
+            return Math.max(leftParentNode.getLayerIndex(), rightParentNode.getLayerIndex()) + 1;
```
This code seems to repeat a lot now. Maybe we extract it?

Christopher-Chianelli committed:
docs: Add documentation for concat in user guide

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,47 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+
+Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
+If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.
+Unlike union of sets, concatenation of lists repeat duplicated elements.
```
```suggestion
Unlike union of sets, concatenation of lists repeats duplicated elements.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,47 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.
+Unlike union of sets, concatenation of lists repeat duplicated elements.
+If the two constraint stream parents share tuples, they will be repeated downstream.
+If this is undesired, use <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,distinct>>.
```
```suggestion
If this is undesired, use the <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,`distinct` building block>>.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,47 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+[source,java,options="nowrap"]
+----
+    private Constraint countCatsAndDogs(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Employee.class)
+         .filter(employee -> employee.hasDog())
+         .concat(constraintFactory.forEach(Employee.class)
+                     .filter(employee -> employee.hasCat()))
```
I am not sure this is a good example. The first question once a smart person sees this is: why not just use a much simple `.filter(hasCat || hasDog)`?

I think we need an example which doesn't have an obvious simpler faster solution.

Christopher-Chianelli replied:
The obvious example would be to implement group by on an outer join, like this:

```java
factory.forEach(A.class)
    .join(B.class, ...)
    .concat(factory.forEach(A.class)
                               .ifNotExists(B.class, ...)
                               .expand(a -> null))
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

Although the example would lead to the question "why no outer join" since that would simplify the code too

```java
factory.forEach(A.class)
    .outerJoin(B.class, ...)
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

The other thing it can be used for is or'ing Joiners, which is unrelated to outer joins

```java
factory.forEach(A.class)
    .join(B.class, equal(a -> a.value, b -> b.value))
    .concat(factory.forEach(A.class)
                               .join(B.class, equal(a -> a.group, b -> b.group)))
    ...
```

triceo replied:
If we're doing concat for the minimal constraint, why not show that? Maybe there's a simplified variant that isn't too much code while still showing the point?

I'm not worried if people start asking about outer join after they see this. First of all, we have a decent answer, this is faster because it doesn't need to do any sort of indexing. Second, if they do, and if there's enough of them, then maybe we eventually do it? It's an opportunity to coax some more real-world use cases out of people, and that's a good thing.

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,47 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+    }
+----
+
+This correctly count cats and dogs when an Employee has both a cat and a dog.
```
```suggestion
This correctly counts cats and dogs when an Employee has both a cat and a dog.
```

triceo replied:
```suggestion
This correctly counts cats and dogs when an Employee has both a cat and a dog.
```

Christopher-Chianelli commented:
TODO: change text in image to use `UNION ALL` instead of `OR` (https://www.techonthenet.com/sql/union_all.php)

Christopher-Chianelli committed:
docs: Update image used for concat to use UNION ALL

Concat is more similar to UNION ALL than OR.

Christopher-Chianelli committed:
docs: Apply suggestions from code review to concat documentation

Co-authored-by: Lukáš Petrovický <lukas@petrovicky.net>

Christopher-Chianelli committed:
docs: Change example used in concat docs

Christopher-Chianelli committed:
chore: Remove references to join in concat comments and refactor getting layer index for binary operations

Christopher-Chianelli committed:
chore: Refactor Nurse Rostering min-max constraint to use concat

This also removes the "minimumNumberOfAssignmentsNoAssignments"
constraint since "minimumAndMaximumNumberOfAssignments" now
handles that case.

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,57 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+[[constraintStreamsConcat]]
+==== Concat
+
+Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
```
```suggestion
The `concat` building block allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,57 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+==== Concat
+
+Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
+If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.
```
```suggestion
If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, `concat` acts like a concatenation of two lists.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,57 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
+If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.
+Unlike union of sets, concatenation of lists repeats duplicated elements.
+If the two constraint stream parents share tuples, they will be repeated downstream.
```
```suggestion
If the two constraint stream parents share tuples, which happens eg. when the streams being concatenated come from the same source of data, the tuples will be repeated downstream.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,57 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+    }
+----
+
+This correctly count the number of shifts each Employee has, *even when the Employee has no shifts*.
```
```suggestion
This correctly counts the number of shifts each Employee has, *even when the Employee has no shifts*.
```

triceo commented on docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc:
```diff
@@ -1511,6 +1511,57 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
+----
+
+This correctly count the number of shifts each Employee has, *even when the Employee has no shifts*.
+If it was implemented without concat like this:
```
```suggestion
If it was implemented without `concat` like this:
```

triceo reviewed (approved):
I added the final set of comments as I re-read the docs.
I'll run my benchmarks overnight to see how it behaves, and then we can merge.

Christopher-Chianelli committed:
docs: Apply suggestions from code review to concat documentation

Co-authored-by: Lukáš Petrovický <lukas@petrovicky.net>

Christopher-Chianelli committed:
fix: Update contents of concat's cloned tuple on update

Previously, concat did not update the content of the cloned tuples
in update. The `groupBy` and `map` building blocks both modify the contents
of their tuples instead of creating new ones. As a result, concat was
passing stale tuples with incorrect content downstream when placed
after a `groupBy` or `map` building block. Now, the contents of the
tuples are updated in update to prevent this issue.

Co-authored-by: Lukáš Petrovický <lukas@petrovicky.net>

triceo commented on core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetBiConcatNode.java:
```diff
@@ -16,4 +16,10 @@ public final class BavetBiConcatNode<A, B> extends AbstractConcatNode<BiTuple<A,
+
+    @Override
+    protected void updateOutTuple(BiTuple<A, B> inTuple, BiTuple<A, B> outTuple) {
+        outTuple.factA = inTuple.factA;
```
Tuples have `updateInDifferent(...)` just for this case.
Turns out it is a common operation, so it was extracted already.

Christopher-Chianelli committed:
chore: Refactor concat to use share code, check if tuple was inserted before updating

Christopher-Chianelli committed:
chore: Inline the updateIfDifferent method in concat nodes
